File: 15_3Sum_1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        for (int cur = 0; cur < nums.length-2; cur++) {
            if (cur !=0 && nums[cur] == nums[cur-1]) {
                continue;
            }else if (nums[cur]+nums[cur+1]+nums[cur+2]>0) {
                break;
            }else if (nums[cur]+nums[nums.length-1]+nums[nums.length-2]<0) {
                continue;
            }
            int i = cur + 1;
            int j = nums.length - 1;
            while (i < j) {
                if (i != cur+1 && nums[i] == nums[i-1]) {
                    i++;
                    continue;
                }
                if (nums[cur] + nums[i] < (-nums[j])) {
                    i++;
                }else if (nums[cur] + nums[i] > (-nums[j])) {
                    j--;
                }else{
                    ans.add(Arrays.asList(nums[cur], nums[i], nums[j]));
                    i++;
                    j--;
                }               
            }
        }
        return ans;
    }
}

File: 15_3Sum_1.py
#! /usr/bin/env python3

class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        ans = list()
        for cur, num in enumerate(nums[:-2]):
            if (cur != 0 and num == nums[cur-1]):
                continue
            elif (num+nums[cur+1]+nums[cur+2] > 0):
                break
            elif (num+nums[len(nums)-1]+nums[len(nums)-2] < 0):
                continue
            
            i = cur + 1
            j = len(nums) - 1
            while(i < j):
                if (i != cur+1 and nums[i] == nums[i-1]):
                    i+=1
                    continue
                
                if (num+nums[i]+nums[j] < 0):
                    i+=1
                elif (num+nums[i]+nums[j] > 0):
                    j-=1
                else:
                    ans.append([num, nums[i], nums[j]])
                    i+=1
                    j-=1
                    
        return ans